### PR TITLE
[TFLite, 16x8] 16x8 Reference kernel for the MEAN operator.

### PR DESCRIPTION
**Pull request by wwwind**
For MobileNet/MobileNetV2 from keras.application we needed 16x8 reference kernel for MEAN operator. This PR adds it. I added versioning and excluded 16x8 tests from running with delegates.
Tests are extended for 16x8 case.

### CODE REVIEW

**Comment by yyoon** [comment]
@jianlijianli Could you take a look?

**Comment by yyoon** [comment]
cc/ @renjie-liu

File: tensorflow/lite/kernels/reduce.cc [hunk 1]
```diff
@@ -397,50 +451,12 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
               GetTensorData<int64_t>(temp_sum)));
       break;
     case kTfLiteInt8: {
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:453]: can we migrate uint8 as well?

**Comment by wwwind** [comment]
Hi @renjie-liu I addressed comments, please take a look. Thanks

File: tensorflow/lite/kernels/reduce.cc [hunk 2]
```diff
@@ -242,7 +245,8 @@ TfLiteStatus PrepareMeanOrSum(TfLiteContext* context, TfLiteNode* node) {
 
   // reduce_mean requires a buffer to store intermediate sum result.
   OpContext op_context(context, node);
-  if (op_context.input->type == kTfLiteInt8) {
+  if (op_context.input->type == kTfLiteInt8 ||
+      op_context.input->type == kTfLiteInt16) {
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:249]: add uint8 here?

File: tensorflow/lite/kernels/internal/reference/reduce.h [hunk 3]
```diff
@@ -320,8 +320,11 @@ inline bool QuantizedMeanOrSum(const T* input_data, int32 input_zero_point,
                                int* temp_index, int* resolved_axis, U* temp_sum,
                                bool compute_sum) {
   const bool uint8_case = std::is_same<T, int8_t>::value;
+  const bool int16_case = std::is_same<T, int16_t>::value;
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/internal/reference/reduce.h:323]: and here?

**Comment by wwwind** [comment]
Hi @renjie-liu Thanks for the review. Your comments are addressed. Please take a look.

File: tensorflow/lite/kernels/reduce.cc [hunk 4]
```diff
@@ -198,6 +198,9 @@ TfLiteStatus InitializeTemporaries(TfLiteContext* context, TfLiteNode* node,
     case kTfLiteInt8:
       temp_sum->type = kTfLiteInt32;
       break;
+    case kTfLiteInt16:
+      temp_sum->type = kTfLiteInt32;
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:202]: nit:

```
 case kTfLiteUInt8:
case kTfLiteInt8:
case kTfLiteInt16:
  temp_sum->type = kTfLiteInt32;
break;
```

**Review by renjie-liu — DISMISSED** [review]

**Review by renjie-liu — DISMISSED** [review]

**Comment by gbaned** [comment]
@wwwind Can you please resolve conflicts? Thanks!

**Comment by wwwind** [comment]
@gbaned done, thanks!

**Comment by wwwind** [comment]
Hi @rthadur Could you re-approve this PR please ? I resolved the conflict. Thanks!

**Review by rthadur — DISMISSED** [review]

File: tensorflow/lite/kernels/reduce.cc [hunk 5]
```diff
@@ -273,6 +274,68 @@ void ResolveAxis(const int* axis_data, int axis_count,
...
+            GetTensorData<integer_type>(op_context.output),
+            op_context.output->params.zero_point,
+            op_context.output->params.scale, op_context.output->dims->data,
+            op_context.output->dims->size, GetTensorData<int>(op_context.axis),
+            num_axis, op_context.params->keep_dims,
+            GetTensorData<int>(temp_index), GetTensorData<int>(resolved_axis),
+            GetTensorData<int>(temp_sum),
+            /*compute_sum=*/false));
+  }
+}
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:337]: this should have a return value here

File: tensorflow/lite/kernels/reduce.cc [hunk 6]
```diff
@@ -400,101 +463,16 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
...
-                GetTensorData<uint8_t>(op_context.output),
-                op_context.output->params.zero_point,
-                op_context.output->params.scale, op_context.output->dims->data,
-                op_context.output->dims->size,
-                GetTensorData<int>(op_context.axis), num_axis,
-                op_context.params->keep_dims, GetTensorData<int>(temp_index),
-                GetTensorData<int>(resolved_axis), GetTensorData<int>(temp_sum),
-                /*compute_sum=*/false));
-      }
+      EvalMeanReferenceOps<uint8_t>(context, op_context, num_axis, data,
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:474]: we should keep the tf_lite_ensure

File: tensorflow/lite/kernels/reduce.cc [hunk 7]
```diff
@@ -400,101 +463,16 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
...
-                GetTensorData<int>(op_context.axis), num_axis,
-                op_context.params->keep_dims, GetTensorData<int>(temp_index),
-                GetTensorData<int>(resolved_axis), GetTensorData<int>(temp_sum),
-                /*compute_sum=*/false));
-      }
+      EvalMeanReferenceOps<int8_t>(context, op_context, num_axis, data,
+                                   temp_index, resolved_axis, temp_sum);
+    } break;
+    case kTfLiteInt16: {
+      EvalMeanReferenceOps<int16_t>(context, op_context, num_axis, data,
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/reduce.cc:470]: keep the tf_lite_esnure

**Comment by gbaned** [comment]
@wwwind Can you please check @renjie-liu's comments and keep us posted. Thanks!

**Comment by gbaned** [comment]
@wwwind Can you please check @renjie-liu's comments and resolve conflicts?. Thanks!

**Comment by rthadur** [comment]
@wwwind did you get chance to address @renjie-liu review comments ?

**Comment by wwwind** [comment]
Hi @rthadur Thanks! I am working on this

**Comment by wwwind** [comment]
Hi @renjie-liu I addressed your comments. Sorry for the delay response. Thanks!

**Review by rthadur — DISMISSED** [review]

File: tensorflow/lite/delegates/nnapi/acceleration_test_list.cc [hunk 8]
```diff
@@ -306,7 +306,9 @@ QuantizeOpTest/INT8,30
 -ConstInt8MeanOpTest.QuantizedDifferentScale
 ConstUint8(Max|Min)OpTest/.+,29
 ConstUint8(Mean)OpTest/.+
-ConstInt8(Mean|Max|Min)OpTest/.+,29
```
Comment (renjie-liu) [inline tensorflow/lite/delegates/nnapi/acceleration_test_list.cc:309]: this is not supported by nnapi right?

File: tensorflow/lite/tools/versioning/op_version.cc [hunk 9]
```diff
@@ -514,6 +514,7 @@ int GetBuiltinOperatorVersion(const OpSignature& op_sig) {
 
     case BuiltinOperator_CONCATENATION:
     case BuiltinOperator_SOFTMAX:
+    case BuiltinOperator_MEAN:
```
Comment (renjie-liu) [inline tensorflow/lite/tools/versioning/op_version.cc:517]: can you update the test as well? 

also the runtime version?

thanks!

**Comment by wwwind** [comment]
Hi @renjie-liu Thanks for the review! I have updated as you suggested. 
I didn't understand where should I add a new test. Could you please clarify ?
Also, I enabled this operator for 16x8 quantization as the interface changes for 16x8 mode quantization have been merged.

**Comment by renjie-liu** [comment]
if you run the tensorflow/lite/tools/versioning/op_version_test
it will fail, so can you update it?

thanks

**Comment by wwwind** [comment]
Hi @renjie-liu I extended this test to check the new version for MEAN operator. Please take a look https://github.com/tensorflow/tensorflow/pull/39988/commits/3c5d55b335e37d893345c44e9ed8c851057c6f70

Thanks!

**Review by renjie-liu — APPROVED** [review]

**Pull request merged**